src/lib/operators/join_hash/multi_predicate_join_evaluator.hpp

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "resolve_type.hpp"
+#include "storage/reference_segment.hpp"
+#include "storage/segment_accessor.hpp"
+#include "storage/table.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+class BaseVoidComparator {
```

Comment:
This file currently still causes us some headache. Is there a better / more modern c++ or hyrise way of doing these comparisons without using void pointers?

Reply:
What's wrong with `std::equal_to`? Also, have you looked at type_comparison.hpp, especially `with_comparator`?

Reply:
Thank you for your suggestion. We will reuse any code we can out of type_comparison.hpp.

The real problem is, however, somewhere else. Please have a look at line 77 following and line 126 following:
To compare two values out of two rows out of two tables we decided to get those values using `void *`. These `void *` will be casted according to the column type of the values two compare. The comparison is done by Comperators. `Comperator<T>` derives from `BaseComperator` which offers the function `bool compare (void* a, void * b)`. `Comperator<T>` implements the compare function by casting `void *` to `T`.

We see the advantage that the ColumnTypes will only have to be resolved once, when creating the comperators. 

Is there a more elegant way using templates which is more secure.

Reply:
We have thought about the `void *` pointers as well. First of all, the `get_void_ptr` method is only used in the `MultiPredicateJoinEvaluator::fulfills_all_predicates` method. This method receives row ids to compare using all additional predicates and needs to have access to the type of the two values to compare for each predicate. A naïve solution would be to adjust the for-loop to something like this:

```cpp
    for (const auto& comparator : _comparators) {
      resolve_column_type(left column of nth join predicate, [](DataType column_type){
        auto lft = dynamic_cast<AbstractAccessor<column_type>>(_left_accessors[col_id][left_row_id.chunk_id]);
        auto rgt = dynamic_cast<AbstractAccessor<column_type>>(_right_accessors[col_id][right_row_id.chunk_id]);

        if (!comparator->compare(lft.access(left_row_id.chunk_offset), rgt.access(right_row_id.chunk_offset)) {
          return false;
        }
      });
      ++col_id;
    }
```

This, however, would be terribly slow, due to the many dynamic casts.

An improved solution might look something like this:

- Comparators have two new instance variables `std::vector<AbstractAccessor<T>> _left_accessors` and `std::vector<AbstractAccessor<T>> _right_accessors` which are supplied during construction. Here is where the accessors are dynamically casted depending on the column type / the column type is used to dynamically resolve the data type.
- Comparators should accept row ids and use the accessors's `.access` method, which returns values of type `T` and circumvents the `void *` pointers. The for-loop then looks something like this:

```cpp
    for (const auto& comparator : _comparators) {
      if (!comparator->compare(left_row_id, right_row_id)) {
        return false;
      }
      ++col_id;
    }
```

This should do the trick and get rid of `void *` pointers once and for all.

Reply:
This seems like a great idea. Let me try that out!